server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -0,0 +1,1192 @@
+      AppenderatorConfig tuningConfig,
+      FireDepartmentMetrics metrics,
+      DataSegmentPusher dataSegmentPusher,
+      ObjectMapper objectMapper,
+      QueryRunnerFactoryConglomerate conglomerate,
+      DataSegmentAnnouncer segmentAnnouncer,
+      ServiceEmitter emitter,
+      ExecutorService queryExecutorService,
+      IndexIO indexIO,
+      IndexMerger indexMerger,
+      Cache cache,
+      CacheConfig cacheConfig
+  )
+  {
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+    this.tuningConfig = Preconditions.checkNotNull(tuningConfig, "tuningConfig");
+    this.metrics = Preconditions.checkNotNull(metrics, "metrics");
+    this.dataSegmentPusher = Preconditions.checkNotNull(dataSegmentPusher, "dataSegmentPusher");
+    this.objectMapper = Preconditions.checkNotNull(objectMapper, "objectMapper");
+    this.conglomerate = conglomerate;
+    this.segmentAnnouncer = segmentAnnouncer;
+    this.emitter = emitter;
+    this.queryExecutorService = queryExecutorService;
+    this.indexIO = indexIO;
+    this.indexMerger = indexMerger;
+    this.cache = cache;
+    this.cacheConfig = cacheConfig;
+
+    if (conglomerate != null) {
+      // If we're not querying (no conglomerate) then it's ok for the other query stuff to be null.
+      // But otherwise, we need them all.
+      Preconditions.checkNotNull(segmentAnnouncer, "segmentAnnouncer");
+      Preconditions.checkNotNull(emitter, "emitter");
+      Preconditions.checkNotNull(queryExecutorService, "queryExecutorService");
+      Preconditions.checkNotNull(cache, "cache");
+      Preconditions.checkNotNull(cacheConfig, "cacheConfig");
+
+      if (!cache.isLocal()) {
+        log.error("Configured cache is not local, caching will not be enabled");
+      }
+    }
+
+    log.info("Creating appenderator for dataSource[%s]", schema.getDataSource());
+  }
+
+  @Override
+  public String getDataSource()
+  {
+    return schema.getDataSource();
+  }
+
+  @Override
+  public Object startJob()
+  {
+    tuningConfig.getBasePersistDirectory().mkdirs();
+    lockBasePersistDirectory();
+    final Object retVal = bootstrapSinksFromDisk();
+    initializeExecutors();
+    resetNextFlush();
+    return retVal;
+  }
+
+  @Override
+  public int add(
+      final SegmentIdentifier identifier,
+      final InputRow row,
+      final Supplier<Committer> committerSupplier
+  ) throws IndexSizeExceededException, SegmentNotWritableException
+  {
+    if (!identifier.getDataSource().equals(schema.getDataSource())) {
+      throw new IAE(
+          "Expected dataSource[%s] but was asked to insert row for dataSource[%s]?!",
+          schema.getDataSource(),
+          identifier.getDataSource()
+      );
+    }
+
+    final Sink sink = getSink(identifier);
+    int sinkRetVal;
+
+    try {
+      sinkRetVal = sink.add(row);
+    }
+    catch (IndexSizeExceededException e) {
+      // Try one more time after swapping, then throw the exception out if it happens again.
```

Comment:
if we swap the index out, and the new, empty index *still* can't hold the row, then the row must be really gigantor and really cannot be indexed.